pkg/ccl/changefeedccl/kvfeed/kv_feed.go

```diff
@@ -337,76 +360,90 @@ func (f *kvFeed) scanIfShould(
+					spansToScan = append(spansToScan, sp)
 				}
 			}
 			if !scanTime.Equal(ev.After.GetModificationTime()) {
-				return errors.AssertionFailedf("found event in shouldScan which did not occur at the scan time %v: %v",
+				return nil, hlc.Timestamp{}, errors.AssertionFailedf(
+					"found event in shouldScan which did not occur at the scan time %v: %v",
 					scanTime, ev)
 			}
 		}
 	} else {
-		return nil
+		return nil, hlc.Timestamp{}, nil
 	}
 
 	// Consume the events up to scanTime.
 	if _, err := f.tableFeed.Pop(ctx, scanTime); err != nil {
-		return err
+		return nil, hlc.Timestamp{}, err
 	}
 
 	// If we have initial checkpoint information specified, filter out
 	// spans which we no longer need to scan.
-	spansToBackfill = filterCheckpointSpans(spansToBackfill, f.checkpoint)
+	spansToBackfill := filterCheckpointSpans(spansToScan, f.checkpoint)
 
 	if (!isInitialScan && f.schemaChangePolicy == changefeedbase.OptSchemaChangePolicyNoBackfill) ||
 		len(spansToBackfill) == 0 {
-		return nil
+		return spansToScan, scanTime, nil
 	}
 
 	if f.onBackfillCallback != nil {
 		defer f.onBackfillCallback()()
 	}
 
-	if err := f.scanner.Scan(ctx, f.writer, physicalConfig{
+	if err := f.scanner.Scan(ctx, f.writer, scanConfig{
 		Spans:     spansToBackfill,
 		Timestamp: scanTime,
 		WithDiff:  !isInitialScan && f.withDiff,
 		Knobs:     f.knobs,
 	}); err != nil {
-		return err
+		return nil, hlc.Timestamp{}, err
 	}
 
-	f.checkpoint = nil
-
-	// NB: We don't update the highwater even though we've technically seen all
-	// events for all spans at the previous highwater.Next(). We choose not to
-	// because doing so would be wrong once we only backfill some tables.
-	return nil
+	// We return entire set of spans (ignoring possible checkpoint) because all of those
+	// spans have been scanned up to and including scanTime.
+	return spansToScan, scanTime, nil
 }
 
-func (f *kvFeed) runUntilTableEvent(
-	ctx context.Context, startFrom hlc.Timestamp,
-) (resolvedUpTo hlc.Timestamp, err error) {
+func (f *kvFeed) runUntilTableEvent(ctx context.Context, frontier *span.Frontier) (err error) {
+	startFrom := frontier.Frontier()
+
 	// Determine whether to request the previous value of each update from
 	// RangeFeed based on whether the `diff` option is specified.
 	if _, err := f.tableFeed.Peek(ctx, startFrom); err != nil {
-		return hlc.Timestamp{}, err
+		return err
 	}
 
 	memBuf := f.bufferFactory()
 	defer func() {
 		err = errors.CombineErrors(err, memBuf.CloseWithReason(ctx, err))
 	}()
 
+	// We have catchup scan checkpoint.  Advance frontier.
+	if startFrom.Less(f.checkpointTimestamp) {
+		for _, s := range f.checkpoint {
+			if _, err := frontier.Forward(s, f.checkpointTimestamp); err != nil {
+				return err
+			}
+		}
+	}
+
+	var stps []kvcoord.SpanTimePair
+	frontier.Entries(func(s roachpb.Span, ts hlc.Timestamp) (done span.OpResult) {
+		stps = append(stps, kvcoord.SpanTimePair{Span: s, TS: ts})
```

Comment:
Instead of forwarding the frontier above, would it not be more correct to do `TS: ts.Next()` here?  Since then we're saying "Our frontier says we've already handled this span up to time `ts`, so we want to start a rangefeed at ts.Next() to get all events after that point".

Reply:
That would be wrong because scanIfShould may have scanned a subset of spans -(i.e. we have feed on 2 tables, but only 1 table had schema change).  That's why scanIfShould now returns the timestamp and the set of spans that it scanned to that timestamp.